core/solver/gcr.cpp

```diff
@@ -0,0 +1,358 @@
+        this->get_system_matrix(),
+        std::shared_ptr<const LinOp>(dense_b, [](const LinOp*) {}), dense_x,
+        residual);
+
+    int total_iter = -1;
+    size_type restart_iter = 0;
+
+    /* Memory movement summary for average iteration with krylov_dim d:
+     * (5d+23+6/d)n * values + matrix/preconditioner stroage
+     * 1x SpMV:                       2n * values + storage
+     * 1x Preconditioner:             2n * values + storage
+     * MGS:                   (5d + 19)n = sum k=0 to d-1 of (10k+24)n/d
+     *       1x dots             2(k+1)n in iteration k (0-based)
+     *       1x scals            2(k+1)n in iteration k (0-based)
+     *       2x axpys            6(k+1)n in iteration k (0-based)
+     *       1x norm2                  n
+     *       1x dot                   2n
+     *       1x sq_norm2               n
+     *       1x step 1 (scal, axpys)  6n
+     *       2x copy                  4n
+     * Restart:                   (6/d)n  (every dth iteration)
+     *       3x copy                  6n
+     */
+    while (true) {
+        ++total_iter;
+        // compute residual norm
+        residual->compute_norm2(residual_norm, reduction_tmp);
+        // Log current iteration
+        this->template log<log::Logger::iteration_complete>(
+            this, total_iter, residual, dense_x, residual_norm);
+        // Check stopping criterion
+        if (stop_criterion->update()
+                .num_iterations(total_iter)
+                .residual(residual)
+                .residual_norm(residual_norm)
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        // If krylov_dim reached, restart with new initial guess
+        if (restart_iter == krylov_dim) {
+            // Restart
+            // p(:, 1) = precon_residual(:, 1)
+            // Ap(:, 1) = A_precon_residual(:, 1)
+            // final_iter_nums = {0, ..., 0}
+            exec->run(gcr::make_restart(precon_residual, A_precon_residual,
+                                        krylov_bases_p, mapped_krylov_bases_Ap,
+                                        final_iter_nums.get_data()));
+            restart_iter = 0;
+        }
+        // compute alpha
+        auto Ap = mapped_krylov_bases_Ap->create_submatrix(
+            span{num_rows * restart_iter, num_rows * (restart_iter + 1)},
+            span{0, num_rhs});
+        auto p = krylov_bases_p->create_submatrix(
+            span{num_rows * restart_iter, num_rows * (restart_iter + 1)},
+            span{0, num_rhs});
+        residual->compute_conj_dot(Ap.get(), tmp_alpha, reduction_tmp);
+        // normalise
+        auto Ap_norm = Ap_norms->create_submatrix(
+            span{restart_iter, restart_iter + 1}, span{0, num_rhs});
+        Ap->compute_squared_norm2(Ap_norm.get(), reduction_tmp);
+
+        // tmp = alpha / Ap_norm
```

Comment:
I guess the nomenclature is consistent to what we have in Ginkgo, but it is somewhat confusing when comparing to Saads book, where `tmp` would be called alpha. I guess the rationale in Ginkgo is usually to use tmp instead of alpha or beta, so that tmp can be reused and serve as both. But here we also have `tmp_alpha` and `tmp_beta`, which makes it hard (for me) to understand what is what.